### ISSUE TITLE
Missing -Wdeprecated warning when user-declared copy assignment operator is defined as deleted

### ISSUE THREAD

**Issue by Quuxplusone** [llvm/llvm-project#44979]
|  |  |
| --- | --- |
| Bugzilla Link | [45634](https://llvm.org/bz45634) |
| Version | trunk |
| OS | All |
| CC | @davidbolvansky,@dwblaikie,@zygoloid |

## Extended Description
```
// https://godbolt.org/z/vmKAwd
struct S {
    int i;
    S& operator=(const S&) = delete;
};
S test(const S& s) { return S(s); }
```

`clang++ -Wdeprecated test.cc` (incorrectly) does not emit a warning for this code.
It will (correctly) emit a warning if you remove the tokens "= delete" from the user-provided declaration.

http://eel.is/c++draft/class.copy.ctor#6 :
> If the class definition does not explicitly declare a copy constructor, a non-explicit one is declared implicitly. If the class definition declares a move constructor or move assignment operator, the implicitly declared copy constructor is defined as deleted; otherwise, it is defined as defaulted ([dcl.fct.def]). The latter case is deprecated if the class has a user-declared copy assignment operator or a user-declared destructor ([depr.impldec]).

S's deleted copy assignment operator counts as a user-declared copy assignment operator; therefore we're in the deprecated case; therefore Clang should warn.

(So says #standardese on Slack, anyway.)

**Comment by Quuxplusone** [comment]
Fixed in abf3ca61e3235681f26d0f527b8e2763dd4c0c62.

**Comment by Kojoley** [comment]
This change affects Boost heavily, we already have [a case in Boost.Proto](https://github.com/boostorg/proto/pull/14#discussion_r310825542) where a struct has to be an aggregate and needs custom copy assignment operator and where the warning could not be fixed without breaking backwards compatibility, now we have a case where a type is also an aggregate where the copy assignment operator is deleted to prevent users from doing strange stuff. Not to mention that a lot of `-Wdeprecated-copy` in conjunction with MSVC C4512 cases were fixed by deleting copy assignment operator.
